File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        
        int leftmax = 0,rightmax = 0;
        int l=0,r= n-1;
        int ans = 0;
        while(l <= r){
            if(height[l] <= height[r]){
                if(height[l] >= leftmax){
                    leftmax = height[l];
                }
                else {
                    ans += (leftmax - height[l]);
                }
                l++;
            }
            else {
                if(height[r] >= rightmax){
                    rightmax = height[r];
                }
                else {
                    ans += (rightmax - height[r]);
                }
                r--;
            }
        }
        return ans;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int low = 0,high = n - 1;
        
        int leftMax = 0,rightMax = 0;
        int ans = 0;
        while(low <= high) {
            
            if(height[low] <= height[high]) {
                if(height[low] > leftMax) {
                    leftMax = height[low];
                }
                else {
                    ans += leftMax - height[low];
                }
                low++;
            }
            else {
                
                if(height[high] > rightMax) {
                    rightMax = height[high];
                }
                else {
                    ans += rightMax - height[high];
                }
                high--;
            }
        }
        return ans;
    }
};